state/store.go

```diff
@@ -212,133 +234,189 @@ func (store dbStore) Bootstrap(state State) error {
 	return store.db.SetSync(stateKey, state.Bytes())
 }
 
-// PruneStates deletes states between the given heights (including from, excluding to). It is not
+// PruneStates deletes states up to the height specified (exclusive). It is not
 // guaranteed to delete all states, since the last checkpointed state and states being pointed to by
-// e.g. `LastHeightChanged` must remain. The state at to must also exist.
-//
-// The from parameter is necessary since we can't do a key scan in a performant way due to the key
-// encoding not preserving ordering: https://github.com/tendermint/tendermint/issues/4567
-// This will cause some old states to be left behind when doing incremental partial prunes,
-// specifically older checkpoints and LastHeightChanged targets.
-func (store dbStore) PruneStates(from int64, to int64) error {
-	if from <= 0 || to <= 0 {
-		return fmt.Errorf("from height %v and to height %v must be greater than 0", from, to)
-	}
-	if from >= to {
-		return fmt.Errorf("from height %v must be lower than to height %v", from, to)
-	}
-	valInfo, err := loadValidatorsInfo(store.db, to)
-	if err != nil {
-		return fmt.Errorf("validators at height %v not found: %w", to, err)
+// e.g. `LastHeightChanged` must remain. The state at retain height must also exist.
+// Pruning is done in ascending order.
+func (store dbStore) PruneStates(retainHeight int64) error {
+	if retainHeight <= 0 {
+		return fmt.Errorf("height %v must be greater than 0", retainHeight)
+	}
+
+	if err := store.pruneValidatorSets(retainHeight); err != nil {
+		return err
 	}
-	paramsInfo, err := store.loadConsensusParamsInfo(to)
+
+	if err := store.pruneConsensusParams(retainHeight); err != nil {
+		return err
+	}
+
+	if err := store.pruneABCIResponses(retainHeight); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (store dbStore) pruneValidatorSets(height int64) error {
+	valInfo, err := loadValidatorsInfo(store.db, height)
 	if err != nil {
-		return fmt.Errorf("consensus params at height %v not found: %w", to, err)
+		return fmt.Errorf("validators at height %v not found: %w", height, err)
 	}
 
-	keepVals := make(map[int64]bool)
+	var (
+		lastRecordedValSetHeight int64
+		lastRecordedValSet       *tmstate.ValidatorsInfo
+	)
+
+	// We will prune up to the validator set at the given "height". As we don't save validator sets every
```

Comment:
Have you tested and vetted this code thoroughly? When I implemented this I believe we had to do this in reverse order since there could be a chain of dependencies that we had to take into account, not just a single height (because different heights point to different lastSaved heights, which again may point to different checkpoints). If we get this wrong it can cause the entire chain to halt and require manual repair, so make extra sure this logic is correct.

Reply:
So I may have missed something that you saw but when I was trying to understand this algorithm, it seemed to me that there was a part in the code that would never be reached. According to the way the last height field is set it shouldn't be possible for there to be a validator set at height x saved where saving a validator set at a higher height y > x has a last height field that is less than the one at x. What this means is that we shouldn't need to keep any more validator sets than the one that the prune height is dependent on.

Reply:
However I would definitely appreciate if you (or someone else) could double check that logic

Reply:
It's been a while since I worked on this, so I'm a bit fuzzy on the details, but remember there were a few gotchas. I can look into this in more depth when you submit the separate PR.